### PR TITLE
Add order handling for fraud outcomes

### LINKED ISSUE #2794: Extract Download button as a separate component
### Description

The **Transactions** page has a Download button to export visible entries in CSV format (#1479). Similar functionality is being added to the **Deposits** page (#2778). Plus, the CSS class names for Transactions page's Download button are already generic (not scoped to the page) — `woocommerce-table__download-button` and `woocommerce-table__download-button__label`.

To avoid code duplication, I propose that we extract the Download button in its own component with its own SCSS and reuse it on the Deposits page (and any other page in the future, eg. Disputes).

### Acceptance criteria

* Extract Download button (export to CSV) as a new component
* Use the new component on Transactions and Deposits pages

### Design

Courtesy @ismaeldcom on #1479:

Annotation on 2021-08-25 at 11-18-50

### PR DESCRIPTION
**Pull request by jessepearson**
Fixes https://github.com/automattic/woocommerce-payments-server/issues/2794

#### Changes proposed in this Pull Request

The purpose of this PR is to add order handling for when fraud outcomes are returning in the meta data of payment intents. The handling includes setting the order status correctly, and also adding/updating meta data in order to have a message displayed in a meta box in the order edit screen.

The `Order_Fraud_And_Risk_Meta_Box` class has been added, which will add the meta box to the order edit page and also display a notice based upon the order's `_wcpay_fraud_outcome_status` meta value.

The larger task here was handling updating the orders based on intent status and the `fraud_outcome` meta data returned in the intent. This required a significant refactor in how the order status was updated from an intent. Intents can be arrays or objects, and due to we need data from the intent, we needed to have a singular place to pull this data from the intent to be used in the order update process.

Most order status update calls have now been piped through the `update_order_status_from_intent` method that was moved from the `WC_Payment_Gateway_WCPay` class to the `WC_Payments_Order_Service` class. The method was also updated to only accept two arguments, `$order` and `$intent`. The new `get_intent_data` method in the order service will then determine if the intent is an array or object and return our needed data.

The status update methods within the `update_order_status_from_intent` have all had their calls to `order_prepared_for_processing` removed so that this can be done and/or checked before any further processing happens. These methods have also all been moved to `private` so that they can only be called through `update_order_status_from_intent`.

All of the status update methods now have their notes generated, and from there we check to make sure the order doesn't already have that note before continuing to process the order. Any early exits in these methods now unlock the orders, as well.

A new method of `get_order` was added to the order service due to the `wc_get_order` function that was being used can return `false` and throw a fatal error. A new exception `Order_Not_Found_Exception` is thrown if an order is not found. This was mainly needed due to some tests have mocked orders and `wc_get_order` attempts to pull an order from the database, and those orders don't actually exist. It's also good practice.

Added new `enqueue_add_fee_breakdown_to_order_notes` method so that fee breakdown notes can be added to all methods that move orders to paid statuses.

Constants were added for the `Fraud_Outcome_Status` possible statuses.

There were several points were redundant or unused code was removed.

#### Testing instructions

The testing for this PR includes a lot of repetition due to UPE and non-UPE have to be tested with auth/capture both and and off. This means going through many steps over and over again, it's simple, but laborious. I recommend using an incognito/private window due to there are session issues with UPE and it's the easiest way to start a new session.

**Test prep**

* There are tests below that are below the _Configurations_, these tests will need to be repeated with each configuration.
* Start listening to Stripe webhooks locally.
* Apply the below patch to the client.
* UPE currently has session issues with intents, so if you get an intent cannot be modified error, increase the quantity of something in your cart by 1 to get around it. You may get this error if you add the same item to your cart, check out, and then add the same item to your cart again.
* Set up _Your Rule_ to use for testing.
  * Update the `wcpay_fraud_protection_settings_active` option in your database to `1`. `wp option update wcpay_fraud_protection_settings_active 1` 
  * Go to Payments > Settings > Fraud Protection, select Advanced, and choose Configure.
  * On the next page, choose a rule you'd like to test with. I found that the last one, _Order Items Threshold_, was the easiest to use.
  * Set the rule to _Block_.

Client patch:
```
diff --git a/includes/class-wc-payment-gateway-wcpay.php b/includes/class-wc-payment-gateway-wcpay.php
index b98e64388..3042002e1 100644
--- a/includes/class-wc-payment-gateway-wcpay.php
+++ b/includes/class-wc-payment-gateway-wcpay.php
@@ -402,7 +402,17 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, [ $this, 'process_admin_options' ] );
 		add_action( 'admin_notices', [ $this, 'display_errors' ], 9999 );
 		add_action( 'woocommerce_order_actions', [ $this, 'add_order_actions' ] );
+		// This is a work around due to you currently get an error when trying to capture charges with UPE enabled.
+		if ( ! get_option( 'woocommerce_order_action_capture_charge_added' ) ) {
 			add_action( 'woocommerce_order_action_capture_charge', [ $this, 'capture_charge' ] );
+			update_option( 'woocommerce_order_action_capture_charge_added', 1 );
+		}
+		add_action(
+			'shutdown',
+			function() {
+				update_option( 'woocommerce_order_action_capture_charge_added', 0 );
+			}
+		);
 		add_action( 'woocommerce_order_action_cancel_authorization', [ $this, 'cancel_authorization' ] );
 
 		add_action( 'wp_ajax_update_order_status', [ $this, 'update_order_status' ] );
diff --git a/woocommerce-payments.php b/woocommerce-payments.php
index 69de822b4..b1dd530a6 100644
--- a/woocommerce-payments.php
+++ b/woocommerce-payments.php
@@ -12,7 +12,7 @@
  * WC tested up to: 7.3.0
  * Requires at least: 5.9
  * Requires PHP: 7.0
- * Version: 5.6.2
+ * Version: 5.7.0
  *
  * @package WooCommerce\Payments
  */
```

**Test to confirm nothing happens if feature is disabled**

1. Update the `wcpay_fraud_protection_settings_active` option in your database to `0`. `wp option update wcpay_fraud_protection_settings_active 0`
1. Add items to your cart to trigger _Your Rule_.
4. Attempt to check out. The order should go through.
6. Go to WooCommerce > Orders in the admin and there should be a new Processing/Completed order, view/edit that order.
7. Confirm there is not a _Fraud & Risk_ meta box showing above the _Order notes_.
1. Update the `wcpay_fraud_protection_settings_active` option in your database to `1` to complete the rest of tests. `wp option update wcpay_fraud_protection_settings_active 1`
2. Refresh the order.
3. Confirm a _Fraud & Risk_ meta box is displayed similar to the below.

Screenshot 2023-03-24 at 1 59 58 PM

**Testing no filtering on non-WCPay transactions**

1. Go to WooCommerce > Settings > Payments > All payment methods (wp-admin/admin.php?page=wc-settings&tab=checkout&section)
2. Enable _Check payments_.
3. Add item to your cart and check out using _Check payments_.
4. Go to WooCommerce > Orders in the admin and view/edit the new order. It should be On Hold.
5. Confirm a _Fraud & Risk_ meta box is displayed similar to the below.

Screenshot 2023-03-24 at 2 03 11 PM

**Configurations**

* UPE enabled and auth/capture enabled.
* UPE enabled and auth/capture disabled.
* UPE disabled and auth/capture enabled.
* UPE disabled and auth/capture disabled.

**Testing steps to repeat**

*Blocked order*
1. Add items to your cart to trigger _Your Rule_.
8. Attempt to check out.
9. You should receive this error: _Error: There's a problem with this payment. Please try again or use a different payment method._
10. Go to WooCommerce > Orders in the admin and there should be a new Pending Payment order, view/edit that order.
11. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below. _Note: that these links are not currently operable._

Screenshot 2023-03-24 at 10 21 13 AM

**Order held for review**

_Note: the below steps can be done by just updating the last name to Review from Block, which will update the existing order instead of creating a new one, as well. The steps assume this is a new session._

1. Add an item to your cart and proceed to checkout.
5. Fill in all information as you normally would, but set your last name as "Review".
6. Check out.
7. Go to WooCommerce > Orders in the admin and there should be a new On hold order, view/edit that order.
8. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below.
9. The links should open a new tab linked to the transaction.

Screenshot 2023-03-24 at 10 28 23 AM

**Capturing order held for review**

1. In the order from the last test, choose _Capture charge_ under _Order actions_, and then click the arrow or _Update_.
3. The order should have updated to Processing/Completed status.
4. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below.
5. The links should open a new tab linked to the transaction.
6. Note: the notice in the meta box will be updated, so it may not match the screenshot.

Screenshot 2023-03-24 at 10 35 29 AM

**Allowed order with auth/capture enabled**

1. Add an item to your cart and proceed to checkout.
2. Fill in all information as you normally would, but set your last name as "Allow".
3. Check out.
4. Go to WooCommerce > Orders in the admin and there should be a new On hold order, view/edit that order.
5. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below.
9. The links should open a new tab linked to the transaction.

Screenshot 2023-03-24 at 11 33 32 AM

**Capturing Allowed order with auth/capture enabled via webhook**

1. Go to Payments > Transactions > Uncaptured
2. The order should appear in the list, click the _Capture_ button.
3. After receiving the confirmation that the charge has been captured, go to the order under WooCommerce > Orders.
4. The order should have updated to Processing/Completed status.
5. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below.
6. The links should open a new tab linked to the transaction.

Screenshot 2023-03-24 at 11 35 05 AM

**Allowed order with auth/capture disabled**

1. Add an item to your cart and proceed to checkout.
2. Fill in all information as you normally would, but set your last name as "Allow".
3. Check out.
4. Go to WooCommerce > Orders in the admin and there should be a new Processing/Complete order, view/edit that order.
5. The order should have a _Fraud & Risk_ meta box and _Order notes_ similar to the below.
6. The links should open a new tab linked to the transaction.

Screenshot 2023-03-24 at 11 57 13 AM

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5782` or branch name `add/2794-fraud-order-details-updates-after-order-service-update` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/2794-fraud-order-details-updates-after-order-service-update) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 631d8d1c59a3355b006c4d787b2a5f83a2a90a96
- Build time: 2023-03-28 14:54:30 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/admin/class-wc-rest-payments-orders-controller.php
Comment (jessepearson) [inline includes/admin/class-wc-rest-payments-orders-controller.php:298]: This code is redundant, as `update_order_status_from_intent` updates the order status.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php]: As stated, the above code is to get around the UPE bug where you cannot capture payments without an error.
https://github.com/Automattic/woocommerce-payments/issues/5780

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php:1407]: This method was moved to WC_Payments_Order_Service.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php:2463]: Above `$currency` variables were unused.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php]: `$charge_id` was unused.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php:2756]: Check was moved to WC_Payments_Order_Service.

File: includes/class-wc-payments-order-service.php
Comment (jessepearson) [inline includes/class-wc-payments-order-service.php:142]: I couldn't find anywhere that set the `_intention_status` to `failed`, so I removed this check.

File: includes/class-wc-payments-order-service.php
Comment (jessepearson) [inline includes/class-wc-payments-order-service.php:441]: `wc_get_order` returns `false` if the order is not found for some reason, so a new method was created to handle getting the order. Calling `get_meta` on `false` causes a fatal.

Discussion about this here: p1679590206525959-slack-C01B8KNUYSW

File: includes/class-wc-payments-order-service.php
Comment (jessepearson) [inline includes/class-wc-payments-order-service.php:1386]: This method was created due to we have intents that can either be objects or arrays, and they are used interchangeably. 

Brought this up to the team: p1677173243169499-slack-C01B8KNUYSW

File: tests/unit/test-class-wc-payment-gateway-wcpay-process-payment.php
Comment (jessepearson) [inline tests/unit/test-class-wc-payment-gateway-wcpay-process-payment.php:225]: This expectation was a duplicate of one just a few lines above, it's actually ignored by PHPUnit.

File: tests/unit/test-class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline tests/unit/test-class-wc-payment-gateway-wcpay.php:2178]: Tests removed since the method was moved to WC_Payments_Order_Service

File: tests/unit/test-class-wc-payments-order-service.php
Comment (jessepearson) [inline tests/unit/test-class-wc-payments-order-service.php:43]: Removed due to we now work with intentions created by `WC_Helper_Intention`

File: includes/class-wc-payment-gateway-wcpay.php
Comment (jessepearson) [inline includes/class-wc-payment-gateway-wcpay.php:2962]: The `get_transaction_url` method was unused.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [inline includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php:87]: Would the description be better phrased as:

> The payment for this order was held for review by your risk filtering. You can review the details and determine whether to capture the charge or cancel the authorization.

This will fall in line with the actions in the _Order actions_ menu.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [inline includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Possible text:

> This transaction was held for review by your risk filters, and then the charge was manually reviewed captured.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [inline includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Possible change to text:

> The payment for this order was blocked by your risk filtering. There is no pending authorization, and the order can be cancelled if you wish.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (tpaksu) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php:87]: Are the merchants more familiar with 'Capture' or 'Approve' in terms of UI? The option on the orders screen actions select says 'Approve', and the transaction details page also contains the button with 'Approve' text in it.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (tpaksu) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Maybe

> This transaction was held for review by your risk filters, and the charge was manually approved after review.

?

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (tpaksu) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: I think we can remove the second sentence completely. No need to indicate, or suggest anything to the merchant. What do you think?

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php:87]: I think they will be more familiar with "Capture charge" and "Cancel authorization". The default text in the Order Edit screen where the meta box is going to appear is what's used here:

```
		$new_actions = [
			'capture_charge'       => __( 'Capture charge', 'woocommerce-payments' ),
			'cancel_authorization' => __( 'Cancel authorization', 'woocommerce-payments' ),
		];
```

https://github.com/Automattic/woocommerce-payments/blob/57a14b1c4214cb3fab2181c55fc9feda8ac50dcd/includes/class-wc-payment-gateway-wcpay.php#L2415-L2418

If we're moving away from this wording and are wanting to use something else, then maybe the options should be updated.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Yours works, I have no idea what I was suggesting. I may have copy, pasted, edited, and accidentally submitted. Maybe I just forgot an "and" between "reviewed captured".

File: includes/constants/class-fraud-outcome-status.php
Comment (tpaksu) [inline includes/constants/class-fraud-outcome-status.php]: I think these statuses mostly match the statuses in https://github.com/Automattic/woocommerce-payments/blob/ee28c1128308c4d0b71a1b7199bbb845379a6c1f/includes/fraud-prevention/models/class-rule.php#L16-L24, except the review results. Do you think it'll be useful for being future-proof if we reference at least `allow`, `review`, and `block` ones to the `Rule` constants?

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (tpaksu) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php:87]: Ok then, what you suggested looks good to me.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Pending Payment orders may be a new thing for them, and if they do not have the Hold Stock setting enabled, they can also block inventory from being able to be purchased. The setting mentioned is under WooCommerce > Settings > Products > Inventory. It enables a periodic job to cancel Pending orders to release their stock back to the purchase pool. 

I think suggesting the order be cancelled would be the way to go. We can also check the Hold Stock setting and change the text according to what is found.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Updated in https://github.com/Automattic/woocommerce-payments/pull/5782/commits/4b4e2aa685151a4577427d1db4159f7d31c53ef3

File: includes/constants/class-fraud-outcome-status.php
Comment (jessepearson) [reply includes/constants/class-fraud-outcome-status.php]: Do you mean setting the constants in `Fraud_Outcome_Status` to something like:

```
const ALLOW = Rule::FRAUD_OUTCOME_ALLOW
```

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: > The payment for this order was blocked by your risk filtering. There is no pending authorization, and the order can be cancelled to reduce any held stock.

What about this?

**Review by tpaksu — APPROVED** [review]
Read the code and made sense, thanks for handling all those cases and simplifying the code with the order service! Also tested each scenario, and all of them work.

|Status|Test|UPE|Auth/Capture|Outcome|
|:----|:----|:----|:----|:----|
|✅|FF disabled|❔|❔|Metabox not showing, order passes through|
|✅|No filtering|❔|❔|Metabox showing as described, order passes through|
|✅|Blocked Order|❌|❌|Order shows pending payment, metabox and note show correct data|
|✅|Blocked Order|❌|✅|No different from auth capture disabled|
|✅|Blocked Order|✅|❌|Order shows pending payment, metabox and note show correct data|
|✅|Blocked Order|✅|✅|No different from auth capture disabled|
|✅|Review Order|❌|❌|Order shows on hold, metabox and note show correct data, capture works, links to correct transaction|
|✅|Review Order|❌|✅|Order shows on hold, metabox and note show correct data, capture works, links to correct transaction|
|✅|Review Order|✅|❌|Order shows on hold, metabox and note show correct data, capture works, links to correct transaction|
|✅|Review Order|✅|✅|Order shows on hold, metabox and note show correct data, capture works, links to correct transaction|
|✅|Allowed order|❌|❌|Order shows completed, metabox and note show correct data|
|✅|Allowed order|❌|✅|Order shows on-hold, metabox and notes show correct data|
|✅|Allowed order|✅|❌|Order shows processing, metabox and note show correct data|
|✅|Allowed order|✅|✅|Order shows on-hold, metabox and notes show correct data|

LGTM! We can ship it!

File: includes/constants/class-fraud-outcome-status.php
Comment (tpaksu) [reply includes/constants/class-fraud-outcome-status.php]: Yep. Only if you're using it to match the fraud outcomes from the server. If those are order statuses, they may need another naming.

File: includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php
Comment (tpaksu) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Works for me.

Comment (jessepearson) [reply includes/fraud-prevention/class-order-fraud-and-risk-meta-box.php]: Updated in https://github.com/Automattic/woocommerce-payments/pull/5782/commits/5f6175bdad4b1117bd46a151c1baf9f38026baa1 , also removed the URL for now.

File: includes/constants/class-fraud-outcome-status.php
Comment (jessepearson) [reply includes/constants/class-fraud-outcome-status.php]: This was updated in https://github.com/Automattic/woocommerce-payments/pull/5782/commits/9f94c3d64200f8ccfd5eba5bc510472af6ff7b36

**Comment by jessepearson** [comment]
This PR has had two significant updates since it was approved that still need to be reviewed before it should be merged:

* https://github.com/Automattic/woocommerce-payments/pull/5782/commits/9f94c3d64200f8ccfd5eba5bc510472af6ff7b36
* https://github.com/Automattic/woocommerce-payments/pull/5782/commits/b30fb2ee2f4f6f6651d5c480db409ccce500d65c

@Automattic/sigma would someone be able to take a look and give their approval?

**Review by tpaksu — APPROVED** [review]
Read the modified code after approval only (2 commits) and still approving again.

**Pull request merged**
